### ISSUE TITLE
noUnusedLocals and noUnusedParameters should flag write-only variables

### ISSUE THREAD

**Issue by andy-ms** [Microsoft/TypeScript#11499]
**TypeScript Version:**  nightly (2.1.0-dev.20161010)

**Code**

(compile with `--noUnusedParameters`)

``` ts
function f(a, b) {
    console.log(arguments)
}
```

**Expected behavior:**

No error

**Actual behavior:** 

```
a.ts(1,12): error TS6133: 'a' is declared but never used.
a.ts(1,15): error TS6133: 'b' is declared but never used.
```

**Comment by RyanCavanaugh** [comment]
You can prefix the names with `_` if you intend to not use them, or create a public signature
```ts
function f(_a, _b) {
    console.log(arguments)
}

function f(_a, _b);
function f() {
    console.log(arguments)
}

```

**Comment by andy-ms** [comment]
While I'm here, should `--noUnusedLocals` flag this?

```ts
function f() {
    let x;
    x = 1;
    return 0;
}
```

**Comment by mhegazy** [comment]
Not per the original design.

**Comment by andy-ms** [comment]
Should `--noUnusedParameters` flag this?

```ts
function f(x: number, y: number) {
    console.log(x)
    if (x > 0)
        f(x - 1, y)
}
```

**Comment by mhegazy** [comment]
no, why? both `x` and `y` are "used".

**Comment by andy-ms** [comment]
These are just patterns I've noticed while looking through the code. I think write-only locals occur often enough to justify detecting, especially for people who prefer imperative-style `let x; if (...) { x = ... } else { x = ... }` to conditional expressions. Presumably we would only have to stop counting writes as references.

**Comment by mhegazy** [comment]
>  Presumably we would only have to stop counting writes as uses, just like we don't count initialization as a use.

this sounds reasonable.
